### PR TITLE
Support pyarrow 14.0.1 and fix vulnerability CVE-2023-47248

### LINKED ISSUE #6396: Issue with pyarrow 14.0.1
See https://github.com/huggingface/datasets-server/pull/2089 for reference

```
from datasets import (Array2D, Dataset, Features)
feature_type = Array2D(shape=(2, 2), dtype="float32")
content = [[0.0, 0.0], [0.0, 0.0]]
features = Features({"col": feature_type})
dataset = Dataset.from_dict({"col": [content]}, features=features)
```

generates

```
/home/slesage/hf/datasets-server/libs/libcommon/.venv/lib/python3.9/site-packages/datasets/features/features.py:648: FutureWarning: pyarrow.PyExtensionType is deprecated and will refuse deserialization by default. Instead, please derive from pyarrow.ExtensionType and implement your own serialization mechanism.
  pa.PyExtensionType.__init__(self, self.storage_dtype)
/home/slesage/hf/datasets-server/libs/libcommon/.venv/lib/python3.9/site-packages/datasets/features/features.py:1661: RuntimeWarning: pickle-based deserialization of pyarrow.PyExtensionType subclasses is disabled by default; if you only ingest trusted data files, you may re-enable this using `pyarrow.PyExtensionType.set_auto_load(True)`.
In the future, Python-defined extension subclasses should derive from pyarrow.ExtensionType (not pyarrow.PyExtensionType) and implement their own serialization mechanism.

  obj = {field.name: generate_from_arrow_type(field.type) for field in pa_schema}
/home/slesage/hf/datasets-server/libs/libcommon/.venv/lib/python3.9/site-packages/datasets/features/features.py:1661: FutureWarning: pyarrow.PyExtensionType is deprecated and will refuse deserialization by default. Instead, please derive from pyarrow.ExtensionType and implement your own serialization mechanism.
  obj = {field.name: generate_from_arrow_type(field.type) for field in pa_schema}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/slesage/hf/datasets-server/libs/libcommon/.venv/lib/python3.9/site-packages/datasets/arrow_dataset.py", line 924, in from_dict
    return cls(pa_table, info=info, split=split)
  File "/home/slesage/hf/datasets-server/libs/libcommon/.venv/lib/python3.9/site-packages/datasets/arrow_dataset.py", line 693, in __init__
    inferred_features = Features.from_arrow_schema(arrow_table.schema)
  File "/home/slesage/hf/datasets-server/libs/libcommon/.venv/lib/python3.9/site-packages/datasets/features/features.py", line 1661, in from_arrow_schema
    obj = {field.name: generate_from_arrow_type(field.type) for field in pa_schema}
  File "/home/slesage/hf/datasets-server/libs/libcommon/.venv/lib/python3.9/site-packages/datasets/features/features.py", line 1661, in <dictcomp>
    obj = {field.name: generate_from_arrow_type(field.type) for field in pa_schema}
  File "/home/slesage/hf/datasets-server/libs/libcommon/.venv/lib/python3.9/site-packages/datasets/features/features.py", line 1381, in generate_from_arrow_type
    return Value(dtype=_arrow_to_datasets_dtype(pa_type))
  File "/home/slesage/hf/datasets-server/libs/libcommon/.venv/lib/python3.9/site-packages/datasets/features/features.py", line 111, in _arrow_to_datasets_dtype
    raise ValueError(f"Arrow type {arrow_type} does not have a datasets dtype equivalent.")
ValueError: Arrow type extension<arrow.py_extension_type<pyarrow.lib.UnknownExtensionType>> does not have a datasets dtype equivalent.
```

### PR DESCRIPTION
**Pull request by albertvillanova**
Support `pyarrow` 14.0.1 and fix vulnerability [CVE-2023-47248](https://github.com/advisories/GHSA-5wvp-7f3h-6wmm).

Fix #6396.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -111,8 +111,8 @@
     # We use numpy>=1.17 to have np.random.Generator (Dataset shuffling)
     "numpy>=1.17",
     # Backend and serialization.
-    # Minimum 8.0.0 to be able to use .to_reader()
-    "pyarrow>=8.0.0",
+    # Minimum 14.0.1 to fix vulnerability CVE-2023-47248
+    "pyarrow>=9.0.0",  # TODO: maximum version allowed by Apache Beam
```
Comment (lhoestq) [inline setup.py:115]: We can require `pyarrow-hotfix<1` and `import pyarrox_hotfix` to fix `CVE-2023-47248` without requiring `14.0.1`

**Comment by albertvillanova** [comment]
The red CI job is unrelated to this PR. It appeared 5 days ago. See:
- https://github.com/huggingface/datasets/pull/6390#pullrequestreview-1721070927
- https://github.com/huggingface/datasets/issues/6406

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -62,9 +62,9 @@ jobs:
       - name: Install dependencies (latest versions)
         if: ${{ matrix.deps_versions == 'deps-latest' }}
         run: pip install --upgrade pyarrow huggingface-hub dill
-      - name: Install depencencies (minimum versions)
+      - name: Install dependencies (minimum versions)
         if: ${{ matrix.deps_versions != 'deps-latest' }}
-        run: pip install pyarrow==8.0.0 huggingface-hub==0.18.0 transformers dill==0.3.1.1
+        run: pip install pyarrow==14.0.1 huggingface-hub==0.18.0 transformers dill==0.3.1.1
```
Comment (lhoestq) [inline .github/workflows/ci.yml:67]:
```suggestion
        run: pip install pyarrow==9.0.0 huggingface-hub==0.18.0 transformers dill==0.3.1.1
```

**Review by lhoestq — APPROVED** [review]
Thanks for the fix !

With this change it won't be possible to load old datasets with ArrayND types saved as Parquet or Arrow anymore, but I don't think we can do anything to avoid that.

**Comment by lhoestq** [comment]
Let's do a new release once this is merged ? cc @mariosasko as well let us know if the fix sounds good to you

**Comment by mariosasko** [comment]
@lhoestq Yes, this sounds good to me!

**Comment by albertvillanova** [comment]
Note I had to add `pa.ExtensionType.__reduce__` because this is used by `copy.deepcopy` when using `.with_format`. See error below.

This method was added in pyarrow-13.0.0: https://github.com/apache/arrow/pull/36170
- We need to re-implement it as long we support lower pyarrow versions

Errors: https://github.com/huggingface/datasets/actions/runs/6861278161/job/18656665772
```
 ____________________________ test_dataset_map[True] ____________________________
[gw1] linux -- Python 3.8.18 /opt/hostedtoolcache/Python/3.8.18/x64/bin/python

>   ???
E   KeyError: 'extension<datasets.features.features.array3dextensiontype<array3dextensiontype>>'

pyarrow/types.pxi:3155: KeyError

During handling of the above exception, another exception occurred:

with_none = True

    @pytest.mark.parametrize("with_none", [False, True])
    def test_dataset_map(with_none):
        ds = datasets.Dataset.from_dict({"path": ["path1", "path2"]})
    
        def process_data(batch):
            batch = {
                "image": [
                    np.array(
                        [
                            [[1, 2, 3], [4, 5, 6], [7, 8, 9]],
                            [[10, 20, 30], [40, 50, 60], [70, 80, 90]],
                            [[100, 200, 300], [400, 500, 600], [700, 800, 900]],
                        ]
                    )
                    for _ in batch["path"]
                ]
            }
            if with_none:
                batch["image"][0] = None
            return batch
    
        features = datasets.Features({"image": Array3D(dtype="int32", shape=(3, 3, 3))})
        processed_ds = ds.map(process_data, batched=True, remove_columns=ds.column_names, features=features)
        assert processed_ds.shape == (2, 1)
>       with processed_ds.with_format("numpy") as pds:

tests/features/test_array_xd.py:459: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/datasets/arrow_dataset.py:2669: in with_format
    dataset = copy.deepcopy(self)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:270: in _reconstruct
    state = deepcopy(state, memo)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:146: in deepcopy
    y = copier(x, memo)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:230: in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:153: in deepcopy
    y = copier(memo)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/datasets/table.py:188: in __deepcopy__
    return _deepcopy(self, memo)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/datasets/table.py:86: in _deepcopy
    setattr(result, k, copy.deepcopy(v, memo))
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:264: in _reconstruct
    y = func(*args)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:263: in <genexpr>
    args = (deepcopy(arg, memo) for arg in args)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:146: in deepcopy
    y = copier(x, memo)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:205: in _deepcopy_list
    append(deepcopy(a, memo))
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:264: in _reconstruct
    y = func(*args)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:263: in <genexpr>
    args = (deepcopy(arg, memo) for arg in args)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:172: in deepcopy
    y = _reconstruct(x, memo, *rv)
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/copy.py:264: in _reconstruct
    y = func(*args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   ValueError: No type alias for extension<datasets.features.features.array3dextensiontype<array3dextensiontype>>

pyarrow/types.pxi:3157: ValueError
```
```
=========================== short test summary info ============================
FAILED tests/test_arrow_dataset.py::BaseDatasetTest::test_class_encode_column_on_disk - ValueError: No type alias for extension<datasets.features.features.array2dextensiontype<array2dextensiontype>>
FAILED tests/test_arrow_dataset.py::BaseDatasetTest::test_dummy_dataset_on_disk - ValueError: No type alias for extension<datasets.features.features.array2dextensiontype<array2dextensiontype>>
FAILED tests/test_arrow_dataset.py::BaseDatasetTest::test_tf_dataset_conversion_in_memory - ValueError: No type alias for extension<datasets.features.features.array2dextensiontype<array2dextensiontype>>
FAILED tests/test_arrow_dataset.py::BaseDatasetTest::test_tf_dataset_conversion_on_disk - ValueError: No type alias for extension<datasets.features.features.array2dextensiontype<array2dextensiontype>>
FAILED tests/test_arrow_dataset.py::BaseDatasetTest::test_tf_dataset_options_in_memory - ValueError: No type alias for extension<datasets.features.features.array2dextensiontype<array2dextensiontype>>
FAILED tests/test_arrow_dataset.py::BaseDatasetTest::test_tf_dataset_options_on_disk - ValueError: No type alias for extension<datasets.features.features.array2dextensiontype<array2dextensiontype>>
FAILED tests/test_arrow_dataset.py::BaseDatasetTest::test_to_csv_on_disk - ValueError: No type alias for extension<datasets.features.features.array2dextensiontype<array2dextensiontype>>
FAILED tests/test_arrow_dataset.py::BaseDatasetTest::test_to_parquet_on_disk - ValueError: No type alias for extension<datasets.features.features.array2dextensiontype<array2dextensiontype>>
FAILED tests/test_arrow_dataset.py::BaseDatasetTest::test_to_sql_on_disk - ValueError: No type alias for extension<datasets.features.features.array2dextensiontype<array2dextensiontype>>
FAILED tests/test_arrow_dataset.py::test_map_cases[True] - ValueError: No type alias for extension<datasets.features.features.array2dextensiontype<array2dextensiontype>>
FAILED tests/test_arrow_dataset.py::test_map_cases[False] - ValueError: No type alias for extension<datasets.features.features.array2dextensiontype<array2dextensiontype>>
FAILED tests/test_arrow_dataset.py::test_map_cases[mix] - ValueError: No type alias for extension<datasets.features.features.array2dextensiontype<array2dextensiontype>>
FAILED tests/features/test_array_xd.py::ArrayXDDynamicTest::test_map_dataset - ValueError: No type alias for extension<datasets.features.features.array3dextensiontype<array3dextensiontype>>
FAILED tests/features/test_array_xd.py::test_dataset_map[False] - ValueError: No type alias for extension<datasets.features.features.array3dextensiontype<array3dextensiontype>>
FAILED tests/features/test_array_xd.py::test_dataset_map[True] - ValueError: No type alias for extension<datasets.features.features.array3dextensiontype<array3dextensiontype>>
===== 15 failed,
```

**Pull request merged**